crates/runix/src/installable.rs

```diff
@@ -1,45 +1,262 @@
+
+        if let Some(start) = start_quote {
+            return Err(ParseInstallableError::UnclosedQuote(
+                s.to_string().split_off(start),
+            ));
+        }
+
+        if !cur.is_empty() {
+            attributes.push_attr(&cur)?;
+        }
+        Ok(attributes)
+    }
+}
 
-        match (split.next(), split.next()) {
-            (Some(flakeref), Some(attr_path)) => Installable {
-                flakeref: flakeref.to_owned(),
-                attr_path: attr_path.to_owned(),
-            },
-            (Some(attr_path), None) => Installable {
-                flakeref: ".".to_owned(),
-                attr_path: attr_path.to_owned(),
-            },
-            _ => unreachable!(),
+impl Display for AttrPath {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some((first, rest)) = self.0.split_first() {
+            write!(f, "{first}")?;
+            for attr in rest {
+                write!(f, ".{attr}")?;
+            }
         }
+        Ok(())
     }
 }
 
-impl ToString for Installable {
-    fn to_string(&self) -> String {
-        self.to_nix()
+impl<A: AsRef<str>> TryFrom<&[A]> for AttrPath {
+    type Error = ParseInstallableError;
+
+    fn try_from(value: &[A]) -> Result<Self, Self::Error> {
+        value
+            .iter()
+            .map(|a| a.as_ref().parse())
+            .collect::<Result<_, _>>()
+            .map(Self)
     }
 }
 
-/// A much simplified FlakeRef representation
+impl<A: AsRef<str>, const N: usize> TryFrom<[A; N]> for AttrPath {
+    type Error = ParseInstallableError;
+
+    fn try_from(value: [A; N]) -> Result<Self, Self::Error> {
+        value.as_slice().try_into()
+    }
+}
+
+/// A validated attribute
 ///
-/// to be replaced by the checked implementation in [crate::flake_ref]
-pub type FlakeRef = String;
+/// Component of an attrpath
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Attribute(String);
+
+impl FromStr for Attribute {
+    type Err = ParseInstallableError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        dbg!(s);
+        if !(VALID_ATTRIBUTE.is_match(s)) {
+            Err(ParseInstallableError::InvalidAttr(s.to_string()))?;
+        }
+        Ok(Attribute(s.to_string()))
+    }
+}
+
+impl Display for Attribute {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.0.contains('.') {
+            write!(f, "\"{}\"", self.0)
+        } else {
+            write!(f, "{}", self.0)
+        }
+    }
+}
+
+impl FromStr for Installable {
+    type Err = ParseInstallableError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.split_once('#') {
+            Some((flakeref, attr_path)) => Ok(Installable {
+                flakeref: flakeref.parse()?,
+                attr_path: attr_path.parse()?,
+            }),
+            None => Ok(Installable {
+                flakeref: s.parse()?,
+                attr_path: "".parse()?,
```

Comment:
Well yeah basically it is a flakeref only but that's supported by nix..
It's about whether we would like to support nix figuring out the "default" attribute itself or not.

Reply:
I also had an `Option` there for a whle but that got a bit messy too
Falls into the same question as where/when to resolve flakeref urls (registry/path/git)..:[#20](https://github.com/flox/runix/issues/20)

For reference, nix will
1. (impurely) turn a given string into a url, https://github.com/NixOS/nix/blob/33aca20616adb872dfab1b3852fe58b948783cd2/src/libexpr/flake/flakeref.cc#L72
2. completely circumventing the flakeref parsing for urls without scheme (https://github.com/NixOS/nix/blob/33aca20616adb872dfab1b3852fe58b948783cd2/src/libexpr/flake/flakeref.cc#L98-L99, https://github.com/NixOS/nix/blob/33aca20616adb872dfab1b3852fe58b948783cd2/src/libexpr/flake/flakeref.cc#L112)  
3. split of the url fragment and validate with regex https://github.com/NixOS/nix/blob/33aca20616adb872dfab1b3852fe58b948783cd2/src/libutil/url.cc
4. then resolve the attrpath from the fragment https://github.com/NixOS/nix/blob/33aca20616adb872dfab1b3852fe58b948783cd2/src/libexpr/attr-path.cc#L9-L32

Reply:
I would just document all of your knowledge in comments - you obviously have a good understanding of how Nix handles a lot of these things informing your design, but since I don't, it's unclear why you're making some of the choices you are

Reply:
Also should this be `AttrPath::default()`?